File: Baekjoon/BaaarkingDog/0x05_Stack/1874.py
# 스택 수열 https://acmicpc.net/problem/1874

####################36784KB	168ms####################
"""
임의의 수열이 주어졌을 때 스택을 이용해 그 수열을 만들 수 있는지 없는지, 
있다면 어떤 순서로 push와 pop 연산을 수행해야 하는지를 계산하는 프로그램을 작성하라
"""

import sys

n = int(sys.stdin.readline())

prev = 0

stack = []
answer = []

for i in range(n):
    curr = int(sys.stdin.readline())

    # 이전에 pop했던 숫자보다 지금 보는 숫자가 크거나 같다면
    if curr >= prev:
        for j in range(prev+1, curr+1):
            stack.append(j)
            answer.append('+')
        # stack에서 pop한 원소가 지금 보낸 숫자와 같다면 - 붙여주고, prev 변수 갱신 
        if curr == stack.pop():
            answer.append('-')
            prev = curr
        
        else:
            print("NO")
            exit(0)
    
    # 이전에 pop했던 숫자보다 지금 보는 숫자가 작다면
    else:
        # stack에서 pop한 원소와 같아야 제대로된 스택 수열임
        if stack.pop() == curr:
            answer.append('-')
        else:
            print("NO")
            exit(0)

print(*answer, sep="\n")

File: Baekjoon/suntouch/1874.js
// 스택 수열 https://www.acmicpc.net/problem/1874
// 스택을 사용해서 수열 nums를 만들 수 있냐!
// 스택에 push하는 순서는 반드시 오름차순!

const path =
  process.platform === "linux" ? "/dev/stdin" : `${__dirname}/1874.txt`;
const fs = require("fs");
const input = fs.readFileSync(path).toString().trim().split("\n");

const count = parseInt(input[0]);
const nums = input.slice(1).map((elem) => parseInt(elem));
const stack = [];
const answers = [];

let lastPopped = 0;

for (let i = 0; i < count; i++) {
  let curr = nums[i];
  // 현재 보고 있는 원소가 마지막에 pop한 요소보다 크거나 같을 때
  if (curr >= lastPopped) {
    for (let j = lastPopped + 1; j < curr + 1; j++) {
      stack.push(j);
      answers.push("+");
    }

    if (curr === stack.pop()) {
      answers.push("-");
      lastPopped = curr; // 사실상 lastPushed와 같음
    } else {
      console.log("NO");
      return;
    }
  }
  //
  else {
    if (curr === stack.pop()) {
      answers.push("-");
      //   lastPopped = curr;
    } else {
      console.log("NO");
      return;
    }
  }
}

console.log(answers.join("\n"));
